### ISSUE TITLE
Covariance / Contravariance Annotations

### ISSUE THREAD

**Issue by Igorbek** [microsoft/TypeScript#1394]
_(It's a question as well as a suggestion)_

_Update_: a proposal #10717

I've supposed that for structural type system as TypeScript is, type variance isn't applicable since type-compatibility is checked by use.

But when I had read @RyanCavanaugh 's [TypeScript 1.4 sneak peek](http://blogs.msdn.com/b/typescript/archive/2014/11/18/what-s-new-in-the-typescript-type-system.aspx) (specifically 'Stricter Generics' section) I realized that there's some lack in this direction by design or implementation. 

I wondered this code is compiled:

``` ts
function push<T>(arr: T[], a: T) { arr.push(a); }
var ns = [1];
push<{}>(ns, "a"); // ok, we added string to number[];
```

More clear code:

``` ts
interface A { a: number; }
interface B extends A { b: number; }
interface C extends B { c: number; }

var a: A, b: B, c: C;

var as: A[], bs: B[];

as.push(a); // ok
as.push(b); // ok, A is contravariant here (any >=A can be pushed) 

bs.push(a); // error, B is contravariant here, so since A<B, A cannot be pushed -- fair 
bs.push(b); // ok
bs.push(c); // ok, C>=B

as = bs;    // ok, covariance used?

as.push(a); // ok, but actually we pushed A to B[]
```

How could `B[]` be assignable to `A[]` if at least on member `push` is not compatible. For `B[].push` it expects parameters of type `B`, but `A[].push` expects `A` and it's valid to call it with `A`.

To illustrate:

``` ts
var fa: (a: A) => void;
var fb: (b: B) => void;

fa(a); fa(b);
fb(a);  // error, as expected
fa = fb;    // no error
fa(a);  // it's fb(a)
```

Do I understand it correctly that is by design?
I don't think it can be called type-safe.

Actually, such restriction that could make `B[]` to be unassignable to `A[]` isn't desirable.
To solve it I suggest to introduce variance on some level (variable/parameter, type?).
#### Syntax

``` ts
var as: A[]; // no variance
var bs: out B[]; // covariant, for "read"
<out A[]>bs; // ok, covariance used
<A[]>bs; // same is before, out should be inferred

(<A[]>bs)[0]; // means, allow to get covariant type
(<A[]>bs).push(a); // means, disallow to pass covariant type

<in A[]>bs; // fails

function push<T>(data: in T[], val: out T): void {
  data.push(val);
}
push(animals, dog); // allowed, T is Animal
push(dogs, animal); // disallow, T can't be inferred

// In opposite
function find<T>(data: out T[], val: out T): bool { ... } // allow only get from data
find(animals, dog); // allowed
find(cats, dog); // allowed T can be inferred as Mammal
```

I'm not sure where variance should be applied - to variable or type?
Looks like it closer to variable it self, so the syntax could be like:

``` ts
var in a: number[];
function<T>(in a: T[], out b: T): { ... }
```
#### Questions to clarification
- inference variance from usage (especially for functions)
- is it applicable for types (like C# uses for interfaces/delegates)
- how to describe variance on member-level (does it need?)
- can variable be `in out` (fixed type)?
- can variable be neither `in` nor `out` (open for upcast/downcast)?

So this topic is a discussion point.

**Comment by rk4n** [comment]
Recently found similar issue, I guess.

```typescript
class Base { n() { return 0; }}
class Derived extends Base { m() { return 1; }}

var base = (b: Base) => b.n();
var derived = (d: Derived) => d.m();

base = derived;

base(new Base()); // TypeError: undefined is not a function (evaluating 'd.m()')
```
Wondering if it could be fixed easily or not?

**Comment by iskiselev** [comment]
Main problem with generic variance is that in TypeScript functions are bivarinat against their arguments. If we only add to function argument signature, that this argument should be strictly contrivance, all generic variance will be calculated automatically. Let's say that it will be achievable with "strict" prefix. Than:
```
interface A{
    A : string;
}

interface B{
    A : string;
    B : string;
}

interface C{
    C : string;
}

interface AOut{
    getItem():A
}

interface BOut{
    getItem():B
}

interface AIn{
    setItem(strict arg : A):void
}

interface BIn{
    setItem(strict arg : B):void
}

interface ConGen<T>
{
    setItem(arg : T):void
}

interface CoGen<T>
{
    getItem():T
}

interface InGen<T>
{
    getItem():T
    setItem(arg : T):void
}

var aOut = <AOut> null;
var bOut = <BOut> null;

aOut =  bOut;
bOut = aOut ; // Same as now - error as function covariant against return type

aOut.getItem = bOut.getItem;
bOut.getItem = aOut.getItem; // Same as now - error as function covariant against return type

var aIn = <AIn> null;
var bIn = <BIn> null;

aIn = bIn; // Error as we marked function argument as contrvariant
bIn = aIn;

aIn.setItem = bIn.setItem; // Error as we marked function argument as contrvariant
bIn.setItem = aIn.setItem;

var aCon = <ConGen<A>> null;
var bCon = <ConGen<B>> null;

aCon = bCon; // Error - generic is calculated as contrvariant
bCon = aCon;

var aCo = <CoGen<A>> null;
var bCo = <CoGen<B>> null;
aCo = bCo;
bCo = aCo; // Same as now, error - generic is calculated as covariant

var aIn = <InGen<A>> null;
var bIn = <InGen<B>> null;
aIn = bIn; // Error - generic is calculated as invariant
bIn = aIn; // Error - generic is calculated as invariant
```

I'll try to do experimental fork with it. Looks like we only need change few lines of code to achieve such behavior.

**Comment by isiahmeadows** [comment]
Found an area where covariance (IIRC) would really help: strongly (and provably) typing the element tree. You could correctly narrow a vdom's children to only the correct types, so something like React could prevent say `<title>` to be a child of a `<div>`, which even though HTML would accept and ignore it, React shouldn't (and doesn't AFAIK). Without covariant types, though, I've already discovered that if you use generics for attributes, TypeScript often requires an explicit parameter to prevent a type error where it wouldn't if the parameter was covariant. A concrete example would be something like this, which doesn't check:

```ts
type Child = // things...

interface Attrs {
  // things...
}

interface VNode<T extends string, A extends Attrs, C extends Child> {
  type: T;
  attrs: A;
  children: C[];
}

export const m<T extends string, C extends Children>(type: T, c?: C | C[]): VNode<T, {}, C>;
export const m<T extends string, A extends Attrs, C extends Children>(type: T, a: A, c?: C | C[]): VNode<T, A, C>;
```

In effect, with covariant types and what exists today (and enough patience - it'd take a long while), you could literally prove the well-formedness of virtual DOM trees and largely correctly type the DOM itself.

**Comment by aleksey-bykov** [comment]
It's rather a simple tweak in the compiler code to turn off covariance for parameters. I tried it once to get a feeling of what it takes to be a happy owner of the code that works right. The biggest pain are the overloaded DOM event handlers lib.d.ts that wont compile after such tweak. I tend to think that using overloading was a poor design choice for event handlers, because effectively they are a bunch of *separate* functions rather than *one* with a base basic signature coupled tons of covariant overloads: 
https://github.com/Microsoft/TypeScript/issues/6102#issuecomment-208119175

**Comment by isiahmeadows** [comment]
@aleksey-bykov Covariance and contravariance usually have to be explicitly specified in type systems with subtyping (Haskell/etc. use type *constraints*, not actual subtyping). It's not easily inferred in general.

**Comment by Igorbek** [comment]
In most cases, it should be easy to infer.

**Comment by isiahmeadows** [comment]
@Igorbek I meant in the case of generics.

@aleksey-bykov Sorry for not fully reading your comment before posting...

I do suspect that adding support for this might allow fixing #6102 later as well. It'd be a breaking change, but after proper generic covariance and contravariance support is added, that would be more likely to be fixable.

**Comment by schotime** [comment]
Really need this feature. Not much point in using inheritance if fuctional-bivariance means that a base class can be passed to a method taking an inherited type.

This should not be possible.

```js
class Animal {}
class Cat extends Animal {}

function run(cat: Cat) {}

run(new Animal());
```

**Comment by Igorbek** [comment]
@schotime it's possible not because of variance. It's possible because Cat and Animal are structurally compatible. If you define `Cat` with a private field, it wouldn't work:
```ts
class Cat extends Animal { private x; }
```
The actual issue with variance and base/derived classes see in [comment](https://github.com/Microsoft/TypeScript/issues/1394#issuecomment-121765250)

**Comment by aleksey-bykov** [comment]
the example that is killing me:
```typescript
interface A {
    x: string;
}

interface B {
    x: string;
    y: string;
}


function copyB(value: B): B {
    return undefined;
}

var values: A[] = [];

values.map(value => copyB(value)) // fails as expected
values.map(copyB); // <-- expected to fail, but it does not
```

https://github.com/Microsoft/TypeScript/issues/3067

**Comment by schotime** [comment]
@Igorbek Apologies. This is a better example.

```
class Animal {
	
}
class Cat extends Animal { private x; }

function run(input: (a: Animal) => void) {
	input(new Animal());
}

function cat(c: Cat) {
	
}

run(cat);
```

**Comment by isiahmeadows** [comment]
@schotime IIUC, those still structurally match.

```ts
type input = (a: Animal) => void;
type cat   = (c: Cat)    => void;
```

The problem is that TypeScript is structural, not nominal. There's currently no way to specify any nominal types other than enums, nominal subtypes of `number`.

So I believe this directly depends on #202. I may be wrong, but it's an educated guess.

**Comment by aleksey-bykov** [comment]
You need to be clearer when you say they structurally match. A callback
requiring a cat cannot be happy when fed by a random animal, say, a whale.
In this sense they are not a near match.

**Comment by isiahmeadows** [comment]
@aleksey-bykov [I stand corrected.](http://www.typescriptlang.org/play/index.html#src=class%20Foo%20%7B%0A%09private%20x%3A%20number%3B%0A%7D%0A%0Aclass%20Bar%20%7B%0A%09private%20x%3A%20number%3B%0A%7D%0A%0Adeclare%20function%20testFoo(foo%3A%20Foo)%3B%0Adeclare%20function%20testBar(bar%3A%20Bar)%3B%0A%0AtestFoo(new%20Bar())%3B%0AtestBar(new%20Foo())%3B%0A)

**Comment by Igorbek** [comment]
Please see my proposal #10717 that addresses the issue. Also [published version](https://github.com/Igorbek/TypeScript-proposals/tree/covariance/covariance).
I'm looking for a feedback from the TypeScript team and the community.

**Comment by aaronla-ms** [comment]
Since no one has mentioned either of the most compelling examples -- arrays and promises. I've added one for good measure:

    var p: Promise<{}> = ...;
    var p2: Promise<string>;
    p2 = p;
    p2.substr(s => s.substr(2)); // runtime error, no method substr on number

Normally you can't assign ``{}`` to ``string`` without a cast, but due to function argument bivarance, you can assign a ``Promise<{}>`` to ``Promise<string>``.

**Comment by isiahmeadows** [comment]
I'd say that's the most compelling reason of all, given the popularity of
promises.

**Comment by yortus** [comment]
@aaronla-ms note that even non-bivariant promise types can be cross-assigned at present. It's another promise issue that masks the bivariance issue.

```ts
var p: Promise<number> = ...;
var p2: Promise<string>;
p2 = p;
p2.then(s => s.substr(2)); // runtime error, no method 'substr' on number
```

**Comment by aaronla-ms** [comment]
@yortus ah yes, I should have clarified that I'm using a simplified promise definition that doesn't have the bug you mentioned:

    interface Promise<T> { 
        then<U>(cb: (value: T) => U | Promise<U>);
    }

Even then, you're able to unsoundly assign Promise<{}> to Promise<string>.

**Comment by RyanCavanaugh** [comment]
FYI this came up in the suggestion backlog slog as parameter bivariance is quite problematic for Promises specifically. We'll be thinking about this more - co/contravariance annotations are the only way out of that particular pickle and it may not be as scary as we think it is.

**Comment by masaeedu** [comment]
Awesome, thanks for considering this! I know its from a couple years ago, but re your comment:

> If function parameters weren't bivariant, Array<Dog> would not be a structural subtype of Array<Animal> due to members like forEach.

`Array<T>.forEach` by itself doesn't seem like it would cause any trouble if `Array` was solely covariant instead of bivariant. E.g. in .NET you have `ForEach<T>` available as an extension method of the covariant `IEnumerable<T>`. 

However it is true that when you arrive at index assignment you are forced to make `Array` itself invariant. You can work around this by having two interfaces, a covariant "iterable" interface and a contravariant "indexable collection" interface like so:

```
// Mark this as covariant
interface Iterable<out T> {
    readonly[index: number]: T;
    length: number;
}

// Mark this as contravariant
interface NumericHash<in T> {
    [index: number]: T;
}

interface Array<T> implements Iterable<T>, NumericHash<T> { ... }
```

Now in the consumer's code:

```
interface Animal { eat(): void }
interface Dog implements Animal { woof(): void; }

let animals: Animal[];
let dogs: Dog[];
let llama: Animal;
let dog: Dog;

// These two are illegal (due to appropriate variance rules)
dogs[0] = llama;
animals.forEach((d: Dog) => d.woof());

// However, these two are fine
animals[0] = dog;
dogs.forEach((a: Animal) => a.eat());
```

**Comment by Igorbek** [comment]
@masaeedu the code
```ts
// Mark this as contravariant
interface NumericHash<in T> {
    [index: number]: T;
}
```
would not really work since the indexer here is for both read and write. Contravariant type would require only write-only side of the index signature which is not supported at the moment.

In my proposal #10717 I'm pointing a way to leave types like `Array` unchanged but leverage use-site variance annotations that would make `Array<in T>` to be a supertype of `Array<T>` where any supertype of `T` can be safely passed (usually that means for write purposes), accordingly `Array<out T>` to be a super type of `Array<T>` where any subtype of `T` can be safely passed (usually for read purposes).

**Comment by masaeedu** [comment]
@Igorbek Ah, okay. That one is just nice to have though, I don't think anyone really cares about using array indexing contravariantly. You could just bake it into the invariant `Array` type.

Your approach is probably the most pragmatic, but I would still really like contra and covariant type annotations to be owned by whoever defines the interface, rather than whoever consumes it. From my pattern of coding at least, having to constantly remember and specify the modifiers where I'm consuming the code would not help me catch bugs. I realize this is subjective however.

**Comment by AlexGalays** [comment]
Another completely unsafe example:

```ts
interface MyComponent<T> {
  items: T[]
  selectedItem: T
  onChange: Handler<T>
}

type Handler<T> = (val: T) => void

function onChange(item: string) {
  console.log(item.slice(2))
}

const comp: MyComponent<string | null> = {
  items: ['a', 'b', 'c'],
  selectedItem: null,
  onChange
}
```

Everything compiles fine but the onChange function can actually get passed a null reference, and that will throw at runtime.

**Comment by PyroVortex** [comment]
With the available tools present in the language, we are nearly to the point where having functions default to contravariant will accomplish most usage scenarios:
```ts
declare func((x: string) => void);
func(() => { }); // Valid
func((x: 'foo') => { }); // Compiler error

// Want a covariant function? Generics will let you write one
declare func2<S extends string>((x: S) => void);
func(() => { }); // Invalid
func2((x: 'foo') => { }); // Valid

interface Array<T> {
    map<U, This1, This2 extends This1>(fn: (this: ThisType, x: T) => U, thisArg?: This2): Array<U>;
    push(x: T);
}
```

**Comment by magnushiie** [comment]
Regarding the argument that co-/contravariance are hard-to-grasp concepts - I think this is irrelevant considering that interface assignment compatibility tracks co-/contravariance even without annotations, and therefore in order for a person to understand what's going on, they need to understand the co-/contravariance (though I think it's covariance vs bivariance currently) distinction anyway.

```typescript
interface Setter<T> {
    set(value: T);
}

let sA1: Setter<{}> = { set() {} };
const sA2: Setter<number> = sA1; 

let sB1: Setter<number> = { set(value: number) {} };
const sB2: Setter<{}> = sB1; // succeeds due to parameter bivariance

interface Getter<T> {
    get(): T;
}

let gA1: Getter<{}> = { get() { return {}; } };
const gA2: Getter<number> = gA1; // Type '{}' is not assignable to type 'number'.

let gB1: Getter<number> = { get() { return 1; } };
const gB2: Setter<{}> = sB1; // succeeds due to covariant returns
```

For a long time, I didn't understand that TypeScript tracks the generic variance, and therefore the distinction between Promise<T> and Iterable<T> was totally confusing - both are reader interfaces, and same variance should apply.

**Comment by nicojs** [comment]
With the [recent addition of `--strictFunctionTypes`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-6.html) co- and contra-variance are now supported on function types without function types. This was a major step up and we had to fix dozens of _actual_ issues in our code (which is awesome!).

Are there any plans to also support co- and contra-variance for other types? (as reported in this issue)? Or nothing on the roadmap?

**Comment by aaronla-ms** [comment]
@nicojs the original proposal seemed to suggest adding co/contra variant annotations as a means of opting out of unsoundness on a per type basis. With `--strictFunctionalTypes`, that variance of a type will now be inferred correctly for other types... my covariant "`interface Stream<T> { next(): T; }`" behaves covariant with respect to `T`, and my contravariant "`interface Bin<T> { put(item: T); }`" behaves contravariant with respect to T.

Could you clarify, are you asking if there are any remaining soundness holes, or asking for an explicit way to declare variance (in addition to the inferred behavior)?

**Comment by Igorbek** [comment]
The proposal initially discussed here and later outlined primarily in #10717 is mostly concerned of use-site covariance annotations. The motivation use cases are still not addressed even with recent enstricten rules.

The simplest example is an array type which can be used both co- and contravariantly. With the proposal   an array of type `Array<out T>` can only be used where type `T` in covariant positions (getters, methods like `pop`). Similarly `Array<in T>` can be used where type `T` in contravariant positions (setters, methods like `push`). Note, in general, `Array<out T>` is not the same as `ReadonlyArray<T>`.

**Comment by isiahmeadows** [comment]
@Igorbek Could you elaborate on how `Array<out T>` would be different from `Readonly<Array<T>>` (not `ReadonlyArray<T>` - I'm not including methods here) or `Array<in T>` from a theoretical `Writeonly<Array<T>>` (assuming some `Writeonly` analogue to the built-in `type Readonly<T> = {+readonly [P in keyof T]: T[P]}`)?

**Comment by Igorbek** [comment]
@isiahmeadows consider `Array`'s methods like `pop`, `reverse`, `shift`, `sort`, and many others. Although they are modifying array and are excluded from `ReadonlyArray<T>`, the type `T` there is in covariant position and therefore is part of `Array<out T>`.

**Comment by isiahmeadows** [comment]
@Igorbek Re-read my comment. You missed my nuance between `Readonly< Array< T > >` and `ReadonlyArray< T >` (spaces here added for emphasis).

**Comment by Igorbek** [comment]
ah, sorry @isiahmeadows I thought you were asking in the context of my previous note about their differences.

In fact, `Readonly<Array<T>>` is even less covariant in respect to `T` than `ReadonlyArray<T>`.
Assuming it is defined as `type Readonly<T> = { readonly [P in keyof T]: T[P]; }`, it only transforms own fields (not even methods) to become read-only.

Actually, TS does not currently make `Readonly<T[]>` be actually read-only:

```ts
function test<T>(a: Readonly<T[]>, v: T) {
    a[0] = v; // no error
}
```

In general, generic type variance has nothing to do with read/write-ability.
A simple test would be:

```ts
interface X<T> {
  value: T; // read-write field with T in covariant position for reads and in contravariant position for writes
  set: (value: T) => void; // read-write field with T in contravariant position
}

type Readonly<X<T>> = { // effectively equivalent to this
  readonly value: T; // this is now covariant in respect to T
  readonly set: (value: T) => void; // read field with T in contravariant position
}

type X<out T> = {
  readonly value T; // same as Readonly
  writeonly set: (value: T) => void; // see the difference
}

type X<in T> = {
  writeonly value T;
  readonly set: (value: T) => void;
}
```

And a final thing, even if in many respects having a readonly version is _mostly_ what you want, a simple `Readonly<T>` cannot deal with multiple type arguments. Like imaging `Processor<TIn, TOut>` is contravariant in respect to `TIn` and covariant in respect to `TOut`, and therefore cannot be expressed with `Readonly`/`Writeonly` as it have mixed type arguments' variance.

**Comment by isiahmeadows** [comment]
Okay, I'd find `a[i] = 0` succeeding when `a: Readonly<number[]>` to be a bug.

**Comment by Igorbek** [comment]
> Okay, I'd find `a[i] = 0` succeeding when `a: Readonly<number[]>` to be a bug.

being fixed in #29435

**Comment by kevinbarabash** [comment]
In the case of arrays, why not prevent aliasing assignments that aren't invariant without an explicit cast?
```
class Animal {}
class Dog extends Animal { bark() { } }
class Cat extends Animal { purr() { } }

const cats: Array<Cat> = [];
const goodAnimals: Array<Animal> = [new Cat, new Cat];
const badAnimals: Array<Animal> = cats;  // why not make this an error?

goodAnimals.push(new Dog);  // okay
badAnimals.push(new Dog);  // hard to prevent, how do you tell goodAnimals from badAnimals?

cats.forEach(cat => cat.purr());  // oops!
```
Given TypeScript's support for generics, when are covariant arrays actually useful?

**Comment by glasser** [comment]
I am very excited to see that this will exist in future versions of TypeScript.

For the time being, if you're looking for a hacky way to ensure contravariance, you can add something like this to a class:

```
export class C<T> {
  protected __forceTToBeContravariant?: (t: T) => void;
```

Note that you can't make this private if you want the constraint to work across TS projects due to https://github.com/microsoft/TypeScript/issues/38953 ; you might think it works in your same-project tests but not once it is encoded in `.d.ts`!
